src/main/java/macrobase/runtime/standalone/scoping/SubSpace.java

```diff
@@ -0,0 +1,341 @@
+package macrobase.runtime.standalone.scoping;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * 
+ * @author xuchu
+ * Represents a subspace in the full dimensional space
+ */
+
+public class SubSpace {
+
+	
+	/**
+	 * A list of ordered dimensions this subspace has, these are scoping dimensions;
+	 */
+	List<Integer> dimensions;
+	
+	
+	/**
+	 * A list of dense units this subspace has on the scoping dimensions
+	 */
+	List<Unit> denseUnits;
+	
+	
+	/**
+	 * The metric dimensions
+	 */
+	List<Integer> metricDimensions;
+	
+	
+	/**
+	 * The following fields are specific to an outlier detection method, right now it is count-based
+	 */
+	Map<Unit,HashMap<Unit,List<Integer>>>  denseUnit2metricUnit2MergedTIDs = new HashMap<Unit,HashMap<Unit,List<Integer>>>();
+	
+	
+	/**
+	 * Create a one dimensional subspace
+	 * @param dimension
+	 */
+	public SubSpace(int dimension){
+		dimensions = new ArrayList<Integer>();
+		dimensions.add(dimension);
+	}
+	
+	public SubSpace(List<Integer> dimensions){
+		this.dimensions = dimensions;
+	}
+	
+	public List<Integer> getDimensions(){
+		return dimensions;
+	}
+	
+	/**
+	 * Return 
+	 * @param other
+	 * @param total
+	 * @param tau
+	 * @return
+	 */
+	public SubSpace join(SubSpace other,int total, double tau){
```

Comment:
The minimum threshold for a Unit considered to be dense. Comment added in the code